### PR TITLE
E2E testing framework

### PR DESCRIPTION
**Pull request by rsevilla87**
### Description

This PR proposes an e2e testing mechanism using the [bats framework](https://github.com/bats-core/bats-core). (You can install it with `dnf install -y bats`)

Some of its benefits:

-  Lazy people friendly. Only bash knowledge is required.
- Code re-usability. Capability of reusing setup() and teardown()  functions across all tests besides of other common functions like checking ES indexing and artifact generation, etc.
- Can run in parallel mode, across test files or within files -> Resulting in a high speed up of the e2e testing of this repo
- Unit-test like implementation.

Working example:
```shell
$ bats --no-parallelize-across-files -j 8 fio.bats 
 ✓ fio-standard
 ✓ fio-bsrange
 ✓ fio-hostpath
 ✓ fio-ocs-cachedrop
4 tests, 0 failures
```

Failing example (after adding an errata to one of the Vegeta  CRs) :
```bash
rsevilla@wonderland ~/ripsaw/e2e (e2e-testing) $ bats --no-parallelize-across-files -j 8 vegeta.bats 
 ✓ vegeta-standard
 ✗ vegeta-hostpath
   (from function `die' in file helpers.bash, line 90,
    in test file vegeta.bats, line 25)
     `check_benchmark 300 || die "Timeout waiting for benchmark/${CR_NAME} to complete"' failed
   benchmark.ripsaw.cloudbulldozer.io/vegeta-benchmark-hostnetwork created
   Waiting for UUID from vegeta-benchmark-hostnetwork
   Timeout waiting for benchmark/vegeta-benchmark-hostnetwork to complete
   Dumping logs at artifacts/vegeta-benchmark-hostnetwork
   Saving log from pod vegeta-0-be7508d6-prrbw in artifacts/vegeta-benchmark-hostnetwork/vegeta-0-be7508d6-prrbw.log
   Saving log from pod vegeta-1-be7508d6-l67lt in artifacts/vegeta-benchmark-hostnetwork/vegeta-1-be7508d6-l67lt.log
   benchmark.ripsaw.cloudbulldozer.io "vegeta-benchmark-hostnetwork" deleted

2 tests, 1 failure
$ ls -lh artifacts/vegeta-benchmark-hostnetwork/
total 1.3M
-rw-rw-r--. 1 rsevilla rsevilla 1.2M Aug  3 14:22 benchmark-controller-manager.log
-rw-rw-r--. 1 rsevilla rsevilla 2.3K Aug  3 14:22 vegeta-0-be7508d6-prrbw.log
-rw-rw-r--. 1 rsevilla rsevilla 2.3K Aug  3 14:22 vegeta-1-be7508d6-l67lt.log
-rw-rw-r--. 1 rsevilla rsevilla 5.9K Aug  3 14:22 vegeta-benchmark-hostnetwork.yaml

```

When manually triggering the github actions workflow we can set a custom image location and select what tests we want to run 
image

Next steps:

- [x] Add remaining test cases.
- [x] Integration with github actions.
- [x] Safe pull request container build mechanism.
- [x] Mechanism to upload artifacts generated (pod logs and benchmark YAML), github workflow run.
- [x] Documentation 

The workflow is executed when in the following conditions:
- Pull request with label "ok to test"
- Commit merged in master branch
- Manually triggered from the "actions" tab of the repository. This option is better described in the documentation

### CODE REVIEW

File: e2e/byowl/byowl-targeted.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    metrics_profile: node-metrics.yml
+  elasticsearch:
+    url: ${ES_SERVER}
+  metadata:
+    collection: true
+    targeted: true
+  workload:
+    name: byowl
+    args:
+      image: "quay.io/jtaleric/uperf:testing"
```
Comment (chentex) [inline e2e/byowl/byowl-targeted.yaml:21]: is this the official image?

File: e2e/fio/fio_bsrange.yaml [hunk 2]
```diff
@@ -28,12 +28,12 @@ spec:
         - randread
         - randrw
       bsrange:
-        - 4KiB-16KiB
+         - 4KiB-16KiB
```
Comment (chentex) [inline e2e/fio/fio_bsrange.yaml:31]: extra space maybe?

**Review by chentex — COMMENTED** [review]
Minor comments.
I cannot really help that much with BATS syntax but did not see anything weird.

File: e2e/byowl/byowl-targeted.yaml (hunk 1 above)
Comment (rsevilla87) [reply e2e/byowl/byowl-targeted.yaml]: Updated it to one from the cloud-bulldozer org

File: e2e/fio/fio_bsrange.yaml (hunk 2 above)
Comment (rsevilla87) [reply e2e/fio/fio_bsrange.yaml]: fixed!

File: README.md [hunk 3]
```diff
@@ -83,6 +83,9 @@ Why did we decide to switch to this? Our operator would implement long running t
 However, long running tasks blocks the Operator, causing us to delete the Operator and re-create the operator to
 un-block it. The benchmarks mentioned above that state `Used` for Reconciliation, no longer have this issue.
 
+# E2E tests
+Benchmark-operator includes a series of end 2 end tests that can be triggered in local. https://github.com/rsevilla87/ripsaw/blob/master/docs/e2e-ci.md#running-in-local
```
Comment (mohit-sheth) [inline README.md:87]: s/rsevilla87/benchmark-operator

**Review by chentex — APPROVED** [review]

**Review by mohit-sheth — APPROVED** [review]
LGTM

**Review by chentex — APPROVED** [review]
LGTM

File: e2e/helpers.bash [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+uuid=""
+suuid=""
+ARTIFACTS_DIR=artifacts
+NAMESPACE=benchmark-operator
+
+
+basic_setup() {
+  export PROMETHEUS_TOKEN=$(oc sa get-token -n openshift-monitoring prometheus-k8s)
+  export ES_SERVER=https://search-perfscale-dev-chmf5l4sh66lvxbnadi4bznl3a.us-west-2.es.amazonaws.com
```
Comment (jtaleric) [inline e2e/helpers.bash:11]: Do we want to hard-set these?

**Review by jtaleric — APPROVED** [review]
small nit, but not worth holding back. LGTM.

File: e2e/helpers.bash (hunk 4 above)
Comment (rsevilla87) [reply e2e/helpers.bash]: Good catch, I have added a new mechanism to customize ES_SERVER by exporting it
The docs now say:

_A preconfigured Elasticsearch endpoint is set by default in the file helpers.bash, however it's possible to set a custom one by exporting the variable ES_SERVER._

**Pull request merged**